### PR TITLE
ZIP-401: weighted random mempool eviction

### LINKED ISSUE #2780: Randomly evict mempool transactions based on their eviction weights
## Motivation

ZIP-401 specifies weighted random mempool transaction evictions:
https://zips.z.cash/zip-0401#specification

### This Ticket Depends On

match by `txid` not `wtxid`:
  - [x] #2819

`serialized transaction size in bytes`:
  - [x] #2778

transaction fee:
  - [x] #2779
    - [x] #2440

### Consensus Rules

> The mempool of a node holds a set of transactions.
> Each transaction has a `cost`, which is an integer defined as:
> `max(serialized transaction size in bytes, 4000)`
>
> Each transaction also has an `eviction weight`, which is `cost + low_fee_penalty`,
> where `low_fee_penalty` is 16000 if the transaction pays a fee less than the conventional fee, otherwise 0. The conventional fee is currently defined as 1000 zatoshis.
>
> Each node also MUST hold a FIFO queue `RecentlyEvicted` of pairs `(txid, time)`, where the `time` indicates when the transaction with the given `txid` was evicted. The `txid` (rather than the `wtxid` ...) is used even for version 5 transactions after activation of NU5.
> 
> On receiving a transaction:
> - If it is in `RecentlyEvicted`, the transaction MUST be dropped.
> - Calculate its `cost`. If the total cost of transactions in the mempool including this one would exceed `mempooltxcostlimit`, then the node MUST repeatedly call `EvictTransaction` (with the new transaction included as a candidate to evict) until the total cost does not exceed `mempooltxcostlimit`.
>
> `EvictTransaction` MUST do the following:
> - Select a random transaction to evict, with probability in direct proportion to `eviction weight`.
> - Add the `txid` and the current `time` to `RecentlyEvicted` ...
> - Remove [the randomly selected transaction] from the mempool.

https://zips.z.cash/zip-0401#specification

### Design

[rand::distributions::weighted::WeightedIndex](https://docs.rs/rand/0.8.4/rand/distributions/weighted/struct.WeightedIndex.html) is a weighted random implementation from a very popular crate.

It supports any collection type with continuous integer indexes. (Like `Vec` or `VecDeque`.)

### PR DESCRIPTION
**Pull request by dconnolly**
## Motivation

We need to do a weighted, randomized eviction of mempool transactions when its total cost gets too high, to comply with ZIP-401.

### Specifications

https://zips.z.cash/zip-040

## Solution

Implements the weighted randomized eviction part of ZIP-401, including calculating the weights of each `VerifiedUnminedTx` and tracking the cost of the verified set in the mempool.

Does not track the evicted set for a limited period of time as specified by ZIP-401, that is being done in #2932 

I got most storage and eviction tests passing by making the max total cost of the mempool for those tests very high, but the tests disabled in d00bd14c9d9a3273c8fb3ad9b2a8a6b513cc8269 need to be fixed.

## Review

This is blocking #2932 and is high priority

### Reviewer Checklist

  - [ ] Code implements Specs and Designs
  - [ ] Tests for Expected Behaviour
  - [ ] Tests for Errors

## Follow Up Work

Re-enable the disabled tests in d00bd14c9d9a3273c8fb3ad9b2a8a6b513cc8269 after #2932

### CODE REVIEW

File: zebra-chain/src/transaction/unmined.rs [hunk 1]
```diff
@@ -17,16 +17,26 @@ use std::{fmt, sync::Arc};
...
 
 use UnminedTxId::*;
 
+/// The minimum cost value for a transaction in the mempool.
+///
+/// Contributes to the randomized, weighted eviction of transactions from the
+/// mempool when it reaches a max size, also based on the total cost.
+///
+/// [ZIP-401]: https://zips.z.cash/zip-0401
+const MEMPOOL_TRANSACTION_COST_THRESHOLD: u32 = 4000;
```
Comment (teor2345) [inline zebra-chain/src/transaction/unmined.rs]: Docs/Audit: Let's quote the definition of the constant here.

That way, we know what we need to change if the ZIP changes.
(Also, it helps the auditors find where we've implemented some rules.)

Can you quote the relevant consensus rules in the docs for this entire PR?

File: zebra-chain/src/transaction/unmined.rs [hunk 2]
```diff
@@ -174,55 +187,117 @@ pub struct UnminedTx {
...
     pub size: usize,
+
+    /// The cost in bytes of the transaction, as defined in [ZIP-401].
+    ///
+    /// A reflection of the work done by the network in processing them (proof
+    /// and signature verification; networking overheads; size of in-memory data
+    /// structures).
+    ///
+    /// [ZIP-401]: https://zips.z.cash/zip-0401
+    pub cost: u32,
```
Comment (teor2345) [inline zebra-chain/src/transaction/unmined.rs]: Design: Since the `cost` is a simple function of the `size`, can we use an `UnminedTx` method instead of another field?

File: zebra-chain/src/transaction/unmined.rs [hunk 3]
```diff
@@ -174,55 +187,117 @@ pub struct UnminedTx {
...
+}
+
+impl fmt::Display for UnminedTx {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("UnminedTx")
+            .field("transaction", &self.transaction)
+            .field("serialized_size", &self.size)
```
Comment (teor2345) [inline zebra-chain/src/transaction/unmined.rs]: Note: we might want to add `cost` to the `Display` impl

File: zebra-chain/src/transaction/unmined.rs [hunk 4]
```diff
@@ -174,55 +187,117 @@ pub struct UnminedTx {
...
 }
 
 // Each of these conversions is implemented slightly differently,
 // to avoid cloning the transaction where possible.
 
 impl From<Transaction> for UnminedTx {
     fn from(transaction: Transaction) -> Self {
+        let size = transaction
+            .zcash_serialized_size()
+            .expect("all transactions have a size");
```
Comment (teor2345) [inline zebra-chain/src/transaction/unmined.rs]: Nitpick: There is one exception to this rule - some generated transactions are structurally invalid, so they won't serialize.

Can you also make this change for every other `size` variable?

```suggestion
            .expect("unexpected serialization failure: all structurally valid transactions have a size");
```

File: zebra-chain/src/transaction/unmined.rs [hunk 5]
```diff
@@ -174,55 +187,117 @@ pub struct UnminedTx {
...
+
         Self {
             id: (&transaction).into(),
-            size: transaction
-                .zcash_serialized_size()
-                .expect("all transactions have a size"),
+            size,
+            cost: std::cmp::max(size as u32, MEMPOOL_TRANSACTION_COST_THRESHOLD),
```
Comment (teor2345) [inline zebra-chain/src/transaction/unmined.rs]: Security: `as` silently truncates overflow. We should panic on overflow instead.

The transaction size check will be implemented in ticket #484 in sprint 20.

Can you also make this change for every other `cost` calculation?

```suggestion
            cost: std::cmp::max(size.try_into().expect("unexpectedly large size: should have already checked the transaction size limit"), MEMPOOL_TRANSACTION_COST_THRESHOLD),
```

File: zebrad/src/components/mempool/storage.rs [hunk 6]
```diff
@@ -114,33 +114,44 @@ pub struct Storage {
...
 }
 
 impl Storage {
-    /// Insert a [`UnminedTx`] into the mempool, caching any rejections.
+    pub(crate) fn new(config: &config::Config) -> Self {
+        Storage {
+            tx_cost_limit: config.tx_cost_limit,
+            ..Default::default()
+        }
+    }
```
Comment (teor2345) [inline zebrad/src/components/mempool/storage.rs]: Refactor: Can we use `Storage::new` in `Mempool`, and make the `Storage` fields private?

(This might be a change for another PR.)

File: zebrad/src/components/mempool/storage.rs [hunk 7]
```diff
@@ -310,6 +323,9 @@ impl Storage {
                 self.tip_rejected_same_effects.insert(txid.mined_id(), e);
             }
             RejectionError::SameEffectsChain(e) => {
+                // TODO: track evicted victims times, removing those older than
+                // config.eviction_memory_time, as well as FIFO more than
+                // MAX_EVICTION_MEMORY_ENTRIES
```
Comment (teor2345) [inline zebrad/src/components/mempool/storage.rs:351]: Future Work: Seems like we might want to switch data structures for the ZIP-401 rejection set.

To get `O(1)` performance, we might need to use:
- `HashMap<ID, Instant>` to make sure IDs are unique and get `O(1)` lookup time, and
- `BTreeSet<(Instant, ID)>` to get `O(1)` expired time removal and length limit enforcement

Did you want to open another ticket for this?

File: zebrad/src/components/mempool/storage/verified_set.rs [hunk 8]
```diff
@@ -87,36 +100,60 @@ impl VerifiedSet {
...
+
+            let weights = self
+                .transactions
+                .iter()
+                .map(|tx| {
+                    let low_fee_penalty = if tx.miner_fee < Amount::try_from(1_000)? {
+                        16_000
+                    } else {
+                        0
+                    };
```
Comment (teor2345) [inline zebrad/src/components/mempool/storage/verified_set.rs]: Refactor: Let's turn `cost + low_fee_penalty` into a method and constants on `VerifiedUnminedTx`.

Can we also:
- quote the relevant consensus rules in their docs
- add it to the `Display` or `Debug` impls (if they already exist, and if they aren't derived)

File: zebrad/src/components/mempool/storage/verified_set.rs [hunk 9]
```diff
@@ -87,36 +100,60 @@ impl VerifiedSet {
...
+                    tx.transaction.cost + low_fee_penalty
+                })
+                .collect();
+
+            let dist = WeightedIndex::new(&weights).unwrap();
```
Comment (teor2345) [inline zebrad/src/components/mempool/storage/verified_set.rs]: Note: collecting and calculating weights is `O(n)`.

But in practice `n` is limited to 20,000 (`mempooltxcostlimit/min(cost)`), so the actual cost shouldn't be too bad.

It's still worth documenting though.

File: zebrad/src/components/mempool/storage/verified_set.rs [hunk 10]
```diff
@@ -144,14 +181,17 @@ impl VerifiedSet {
...
             .remove(transaction_index)
             .expect("invalid transaction index");
 
-        self.transactions_serialized_size -= removed_tx.size;
-        self.remove_outputs(&removed_tx);
+        self.transactions_serialized_size -= removed_tx.transaction.size;
+        self.remove_outputs(&removed_tx.transaction);
+        self.transactions_serialized_size -= removed_tx.transaction.size;
+        self.total_cost -= removed_tx.transaction.cost;
+        self.remove_outputs(&removed_tx.transaction);
```
Comment (teor2345) [inline zebrad/src/components/mempool/storage/verified_set.rs]: Bug: some of these lines are duplicated.

I'd expect this to cause a panic or a test failure. If it doesn't, let's make sure we write tests for it.

File: zebrad/src/components/mempool/storage/verified_set.rs [hunk 11]
```diff
@@ -225,5 +265,6 @@ impl VerifiedSet {
             "zcash.mempool.size.bytes",
             self.transactions_serialized_size as _
         );
+        metrics::gauge!("zcash.mempool.total_cost.bytes", self.total_cost as _);
```
Comment (teor2345) [inline zebrad/src/components/mempool/storage/verified_set.rs]: Note: `total_` has a special meaning to Prometheus.

> `total` (for a unit-less accumulating count)

https://prometheus.io/docs/practices/naming/

```suggestion
        metrics::gauge!("zcash.mempool.cost.bytes", self.total_cost as _);
```

File: zebrad/src/components/mempool/storage/verified_set.rs [hunk 12]
```diff
@@ -87,36 +100,60 @@ impl VerifiedSet {
...
+                    } else {
+                        0
+                    };
+                    tx.transaction.cost + low_fee_penalty
+                })
+                .collect();
+
+            let dist = WeightedIndex::new(&weights).unwrap();
+
+            Some(self.remove(dist.sample(&mut thread_rng())))
```
Comment (teor2345) [inline zebrad/src/components/mempool/storage/verified_set.rs:153]: This is nice and simple, thanks! 🎉

**Review by teor2345 — COMMENTED** [review]
This looks like a good start!
I'm glad we could do weighted eviction in a simple way.

We should talk about how we can get ZIP-401 done over the next few days in our sprint meeting.

File: zebrad/src/components/mempool/storage.rs (hunk 7 above)
Comment (teor2345) [reply zebrad/src/components/mempool/storage.rs:351]: Or we could use an ordered map crate.

This one could be useful:
https://crates.io/crates/total-order-multi-map

Although we don't really need a multi-map.

File: zebra-chain/src/transaction/unmined.rs (hunk 3 above)
Comment (dconnolly) [reply zebra-chain/src/transaction/unmined.rs]: Do we want this if `cost()` is now a function?

File: zebra-chain/src/transaction/unmined.rs (hunk 4 above)
Comment (dconnolly) [reply zebra-chain/src/transaction/unmined.rs]: Resolved in a9e02fae

File: zebra-chain/src/transaction/unmined.rs (hunk 3 above)
Comment (teor2345) [reply zebra-chain/src/transaction/unmined.rs]: I don't mind either way, so I'm just going to resolve this.

File: zebrad/src/components/mempool/storage/verified_set.rs [hunk 13]
```diff
@@ -24,12 +26,15 @@ use super::super::SameEffectsTipRejectionError;
...
     /// The set of verified transactions in the mempool.
-    transactions: VecDeque<UnminedTx>,
+    transactions: VecDeque<VerifiedUnminedTx>,
 
     /// The total size of the transactions in the mempool if they were
     /// serialized.
     transactions_serialized_size: usize,
 
+    /// The total cost of the verified transactons in the set.
+    total_cost: u32,
```
Comment (teor2345) [inline zebrad/src/components/mempool/storage/verified_set.rs]: Security Risk: Let's match the type of `cost()` here.

Adding `u64`s to a `u32` total increases the risk of silent truncation in release code.
(That's unlikely due to our size limits and parsing algorithm, but let's just be safe!)

```suggestion
    total_cost: u64,
```

File: zebra-chain/src/transaction/unmined.rs (hunk 1 above)
Comment (dconnolly) [reply zebra-chain/src/transaction/unmined.rs]: Resolved in 7600d2eb517ce455b5dd39c78b47800ed434b086

File: zebrad/src/components/mempool/storage.rs (hunk 6 above)
Comment (dconnolly) [reply zebrad/src/components/mempool/storage.rs]: Done in 85f0e68554e69e0fd63a07d010853f23100b2234

File: zebrad/src/components/mempool/storage.rs (hunk 7 above)
Comment (conradoplg) [reply zebrad/src/components/mempool/storage.rs:351]: This will be https://github.com/ZcashFoundation/zebra/issues/2759, right?

File: zebra-chain/src/transaction/unmined.rs (hunk 1 above)
Comment (teor2345) [reply zebra-chain/src/transaction/unmined.rs]: I think you might have forgotten to push this commit?

File: zebrad/src/components/mempool/storage.rs (hunk 7 above)
Comment (teor2345) [reply zebrad/src/components/mempool/storage.rs:351]: Yes, this is resolved by PR #2932.

File: zebrad/src/components/mempool/storage/verified_set.rs (hunk 8 above)
Comment (dconnolly) [reply zebrad/src/components/mempool/storage/verified_set.rs]: Partially done in 87e3ddf99c4f0b6dfd69656fb5cef637e88ced58

File: zebra-chain/src/transaction/unmined.rs (hunk 1 above)
Comment (dconnolly) [reply zebra-chain/src/transaction/unmined.rs]: Yeah not pushed /yet/

File: zebrad/src/components/mempool/storage/verified_set.rs (hunk 13 above)
Comment (dconnolly) [reply zebrad/src/components/mempool/storage/verified_set.rs]: I think addressed in 6fe52368, all `cost` values are now `u64`

File: zebrad/src/components/mempool/storage/verified_set.rs (hunk 9 above)
Comment (dconnolly) [reply zebrad/src/components/mempool/storage/verified_set.rs]: Noted in 0d3c7a24

File: zebra-chain/src/transaction/unmined.rs (hunk 1 above)
Comment (dconnolly) [reply zebra-chain/src/transaction/unmined.rs]: Oof rebase changed these, 79b41e90a49e3733ab317c3e5319d93fa9a33962

Comment (dconnolly) [reply zebra-chain/src/transaction/unmined.rs]: e0d23c6735f5ae288f6db10b68f454dc6eaf4b47

File: zebrad/src/components/mempool/storage/verified_set.rs (hunk 8 above)
Comment (dconnolly) [reply zebrad/src/components/mempool/storage/verified_set.rs]: Done,  consensus rules documented, `VerifiedUnminedTx` derives `Debug` automatically

**Comment by dconnolly** [comment]
Fixing tests

**Comment by dconnolly** [comment]
> Fixing tests

I may remove some tests that are currently failing and add them back after #2889 is merged

**Comment by dconnolly** [comment]
@teor2345 @conradoplg  I got most storage and eviction tests passing by making the max total cost of the mempool for those tests very high, but the tests disabled in d00bd14c9d9a3273c8fb3ad9b2a8a6b513cc8269 need to be fixed.

**Comment by teor2345** [comment]
> @teor2345 @conradoplg I got most storage and eviction tests passing by making the max total cost of the mempool for those tests very high, but the tests disabled in [d00bd14](https://github.com/ZcashFoundation/zebra/commit/d00bd14c9d9a3273c8fb3ad9b2a8a6b513cc8269) need to be fixed.

Once you've got everything passing, let's open a ticket for the tests we want to add back later.

File: zebrad/src/components/mempool/storage/tests/vectors.rs [hunk 14]
```diff
@@ -71,79 +77,6 @@ fn mempool_storage_crud_same_effects_mainnet() {
     assert!(!storage.contains_transaction_exact(&unmined_tx.transaction.id));
 }
 
-#[test]
```
Comment (conradoplg) [inline zebrad/src/components/mempool/storage/tests/vectors.rs:74]: I suggest just commenting out the `#[test]` (which will probably require adding `#[allow(dead_code)]`) instead of deleting the code, since that would reduce the changes of conflicts in the future, in case it takes a while to fix the tests

(Unless they're not compiling?)

File: zebra-chain/src/transaction/unmined.rs [hunk 15]
```diff
@@ -29,6 +29,24 @@ use crate::{
...
 use UnminedTxId::*;
 
+/// The minimum cost value for a transaction in the mempool.
+///
+/// Contributes to the randomized, weighted eviction of transactions from the
+/// mempool when it reaches a max size, also based on the total cost.
+///
+/// > Each transaction has a cost, which is an integer defined as:
+/// >
+/// >     max(serialized transaction size in bytes, 4000)
```
Comment (conradoplg) [inline zebra-chain/src/transaction/unmined.rs]: rustdoc thinks this is a doctest, not sure if there's a better way to fix this

```suggestion
/// > max(serialized transaction size in bytes, 4000)
```

File: zebra-chain/src/transaction/unmined.rs [hunk 16]
```diff
@@ -271,4 +297,45 @@ impl VerifiedUnminedTx {
...
+
+    /// The cost in bytes of the transaction, as defined in [ZIP-401].
+    ///
+    /// A reflection of the work done by the network in processing them (proof
+    /// and signature verification; networking overheads; size of in-memory data
+    /// structures).
+    ///
+    /// > Each transaction has a cost, which is an integer defined as:
+    /// >
+    /// >     max(serialized transaction size in bytes, 4000)
```
Comment (conradoplg) [inline zebra-chain/src/transaction/unmined.rs]: rustdoc thinks this is a doctest, not sure if there's a better way to fix this

```suggestion
    /// > max(serialized transaction size in bytes, 4000)
```

**Review by conradoplg — DISMISSED** [review]
Looks good! I left a suggestion and a possible fix for the doctest compile failure

File: zebra-chain/src/transaction/unmined.rs (hunk 16 above)
Comment (teor2345) [reply zebra-chain/src/transaction/unmined.rs]: Using a block quote and "text" tells Rust that it's text, not code:
https://doc.rust-lang.org/rustdoc/documentation-tests.html#attributes

File: zebrad/src/components/mempool/storage/tests/vectors.rs (hunk 14 above)
Comment (teor2345) [reply zebrad/src/components/mempool/storage/tests/vectors.rs:74]: I don't mind what we do here, but we should open a ticket to make sure we either delete or fix the tests.
(Leaving unused tests in the code can be really confusing.)

File: zebra-chain/src/transaction/unmined.rs [hunk 17]
```diff
@@ -271,4 +297,45 @@ impl VerifiedUnminedTx {
...
+    /// > a fee less than the conventional fee, otherwise 0. The conventional fee
+    /// > is currently defined as 1000 zatoshis
+    ///
+    /// [ZIP-401]: https://zips.z.cash/zip-0401
+    pub fn eviction_weight(self) -> u64 {
+        let conventional_fee = 1000;
+
+        let low_fee_penalty = if u64::from(self.miner_fee) < conventional_fee {
+            16_000
+        } else {
```
Comment (teor2345) [inline zebra-chain/src/transaction/unmined.rs:335]: Code Cleanup: We could turn these integer literals into constants.

File: zebrad/src/components/mempool/storage.rs [hunk 18]
```diff
@@ -136,6 +140,13 @@ impl Drop for Storage {
 }
 
 impl Storage {
+    #[allow(clippy::field_reassign_with_default)]
+    pub(crate) fn new(config: &config::Config) -> Self {
+        let mut default: Storage = Default::default();
+        default.tx_cost_limit = config.tx_cost_limit;
```
Comment (teor2345) [inline zebrad/src/components/mempool/storage.rs:146]: Code Cleanup: We could delete the `Default` impl, so it's impossible to create a `Storage` without applying a `Config`.

I think PR #2932 stops using `Storage::default`, but doesn't delete the `Default` impl.

File: zebrad/src/components/mempool/storage.rs [hunk 19]
```diff
@@ -172,20 +183,22 @@ impl Storage {
         }
 
         // Once inserted, we evict transactions over the pool size limit.
```
Comment (teor2345) [inline zebrad/src/components/mempool/storage.rs:185]: Docs/Audit Cleanup: Let's quote the ZIP-401 rule here, as part of ticket #2958.

File: zebrad/src/components/mempool/storage.rs [hunk 20]
```diff
@@ -172,20 +183,22 @@ impl Storage {
...
-        while self.verified.transaction_count() > MEMPOOL_SIZE {
-            let evicted_tx = self
+        while self.verified.transaction_count() > MEMPOOL_SIZE
```
Comment (teor2345) [inline zebrad/src/components/mempool/storage.rs:186]: I opened #2958 to remove `MEMPOOL_SIZE`, because it's not actually part of ZIP-401.

File: zebrad/src/components/mempool/storage/tests/vectors.rs (hunk 14 above)
Comment (teor2345) [reply zebrad/src/components/mempool/storage/tests/vectors.rs:74]: The way we disable tests isn't a blocker for merging this PR, and we're on a tight schedule.

I opened #2959 to fix these tests.

File: zebrad/src/components/mempool/tests/prop.rs [hunk 21]
```diff
@@ -84,94 +81,6 @@ proptest! {
         })?;
     }
 
-    /// Test if the mempool storage is cleared on multiple chain resets.
```
Comment (teor2345) [inline zebrad/src/components/mempool/tests/prop.rs:87]: Was this test deleted when merging an updated `main` branch?

I opened #2959 to fix these tests.

File: zebrad/src/components/mempool/tests/vector.rs [hunk 22]
```diff
@@ -25,196 +25,6 @@ type StateService = Buffer<BoxService<zs::Request, zs::Response, zs::BoxError>,
 /// A [`MockService`] representing the Zebra transaction verifier service.
 type MockTxVerifier = MockService<tx::Request, tx::Response, PanicAssertion, TransactionError>;
 
-#[tokio::test]
-async fn mempool_service_basic() -> Result<(), Report> {
```
Comment (teor2345) [inline zebrad/src/components/mempool/tests/vector.rs:29]: I opened #2959 to fix these tests.

**Review by teor2345 — APPROVED** [review]
I can't see any blockers for this PR, so I'm going to auto-merge it after CI passes.

We're on a tight deadline, so I've split the remaining mempool work into separate tickets.

File: zebrad/src/components/mempool/storage.rs (hunk 18 above)
Comment (conradoplg) [reply zebrad/src/components/mempool/storage.rs:146]: I did this in https://github.com/ZcashFoundation/zebra/pull/2932
